public/app/features/scenes/services/UrlSyncManager.ts

```diff
@@ -1,35 +1,176 @@
-      const urlUpdate = changedObject.getUrlState();
-      locationService.partial(urlUpdate, true);
+    if (changedObject.urlSync) {
+      const newUrlState = changedObject.urlSync.getUrlState(payload.newState);
+      const prevUrlState = changedObject.urlSync.getUrlState(payload.prevState);
+
+      const searchParams = locationService.getSearch();
+      const mappedUpdated: SceneObjectUrlValues = {};
+
+      this.urlKeyMapper.rebuldIndex(this.sceneRoot);
+
+      for (const [key, newUrlValue] of Object.entries(newUrlState)) {
+        const uniqueKey = this.urlKeyMapper.getUniqueKey(key, changedObject);
+        const currentUrlValue = searchParams.getAll(uniqueKey);
+
+        if (!isUrlValueEqual(currentUrlValue, newUrlValue)) {
+          mappedUpdated[uniqueKey] = newUrlValue;
+
+          // Remember the initial state so we can go back to it
+          if (!this.initialStates.has(uniqueKey) && prevUrlState[key] !== undefined) {
+            this.initialStates.set(uniqueKey, prevUrlState[key]);
+          }
+        }
+      }
+
+      if (Object.keys(mappedUpdated).length > 0) {
+        locationService.partial(mappedUpdated, false);
+      }
     }
   };
 
   public cleanUp() {
     this.stateChangeSub.unsubscribe();
     this.locationListenerUnsub();
   }
+
+  private syncSceneStateFromUrl(sceneObject: SceneObject, urlParams: URLSearchParams) {
+    if (sceneObject.urlSync) {
+      const urlState: SceneObjectUrlValues = {};
+      const currentState = sceneObject.urlSync.getUrlState(sceneObject.state);
+
+      for (const key of sceneObject.urlSync.getKeys()) {
+        const uniqueKey = this.urlKeyMapper.getUniqueKey(key, sceneObject);
+        const newValue = urlParams.getAll(uniqueKey);
+        const currentValue = currentState[key];
+
+        if (isUrlValueEqual(newValue, currentValue)) {
+          continue;
+        }
+
+        if (newValue.length > 0) {
+          if (Array.isArray(currentValue)) {
+            urlState[key] = newValue;
+          } else {
+            urlState[key] = newValue[0];
+          }
+
+          // Remember the initial state so we can go back to it
+          if (!this.initialStates.has(uniqueKey) && currentValue !== undefined) {
+            this.initialStates.set(uniqueKey, currentValue);
+          }
+        } else {
+          const initialValue = this.initialStates.get(uniqueKey);
+          if (initialValue !== undefined) {
+            urlState[key] = initialValue;
+          }
+        }
+      }
+
+      if (Object.keys(urlState).length > 0) {
+        sceneObject.urlSync.updateFromUrl(urlState);
+      }
+    }
+
+    forEachSceneObjectInState(sceneObject.state, (obj) => this.syncSceneStateFromUrl(obj, urlParams));
+  }
+}
+
+interface SceneObjectWithDepth {
+  sceneObject: SceneObject;
+  depth: number;
+}
+class UniqueUrlKeyMapper {
+  private index = new Map<string, SceneObjectWithDepth[]>();
+
+  public getUniqueKey(key: string, obj: SceneObject) {
```

Comment:
The only thing that's slightly concerning to me - but maybe I'm going too far into the future - is that the address is dependent on the depth that can and will change if the object is moved to a different depth level within the scene. That's why when I was thinking about URL sync I considered addressing URL keys with actual object keys so that the URL is layout structure independent. I guess it's more stable for dynamic scenes scenarios, but resolving the state from URL is then impossible as the key is generated on object creation :/

I'm primarily thinking of a scenario when the scene is dynamically modified - meaning that the objects can potentially be added above the current object, pushing its index down and making the URL out of sync. But - maybe this is not an issue as the index is rebuilt on every state change?

Reply:
@dprokop I agree, that in a world where you can move scene objects around we might need something more static, using key by default long is problematic though, as they are long guids and would make the URL really long and messy. Having the url be readable and short is important (and compatible with the current from&to for dashboards/scenes with single time range, which will be the most common) 

Think we can explore other options here, The UrlSyncManager could either have different modes/options for how to handle this, or we could add options to the SceneObjectUrlSyncHandler interface. 

But think the current behavior implemented in this PR as a simple default behavior is a good starting point until we encounter issues with it. 

We might need to introduce a panelId numerical id to VizPanel to support old dashboards, maybe we extend this to all scene objects (give them a unique numerical id instead of string guid), then that would be easier to use in the url as a sufix to make keys unique

Reply:
> But think the current behavior implemented in this PR as a simple default behavior is a good starting point until we encounter issues with it.

Yes, definitely agree, just needed a brain dump :D